src/brod_group_controller.erl

```diff
@@ -0,0 +1,889 @@
+        , offset_commit_policy           :: offset_commit_policy()
+        , offset_commit_interval_seconds :: pos_integer()
+        }).
+
+-define(IS_LEADER(S), (S#state.leaderId =:= S#state.memberId)).
+
+%%%_* APIs =====================================================================
+
+%% @doc To be called by group subscriber.
+%% Client:  ClientId (or pid, but not recommended)
+%% GroupId: Predefined globally unique (in a kafka cluster) binary string.
+%% Topics:  Predefined set of topic names in the group.
+%% Config:  The group controller configs in a proplist, possible entries:
+%%  - partition_assignment_strategy  (optional, default = roundrobin)
+%%      roudrobin: Take all topic-offset (sorted [{TopicName, Partition}] list)
+%%                 assign one to each member in a roundrobin fashion.
+%%      TODO: support sticky assignments
+%%  - session_timeout_seconds (optional, default = 10)
+%%      Time in seconds for the group coordinator broker to consider a member
+%%      'down' if no heartbeat or any kind of requests received from a broker
+%%      in the past N seconds.
+%%      A group member may also consider the coordinator broker 'down' if no
+%%      heartbeat response response received in the past N seconds.
+%%  - heartbeat_rate_seconds (optional, default = 2)
+%%      Time in seconds for the member to 'ping' the group coordinator.
+%%      OBS: Care should be taken when picking the number, on one hand, we do
+%%           not want to flush the broker with requests if we set it too low,
+%%           on the other hand, if set it too high, it may take too long for
+%%           the members to realise status changes of the group such as
+%%           assignment rebalacing or group coordinator switchover etc.
+%%  - max_rejoin_attempts (optional, default = 5)
+%%      Maximum number of times allowd for a member to re-join the group.
+%%      The gen_server will stop if it reached the maximum number of retres.
+%%      OBS: 'let it crash' may not be the optimal strategy here because
+%%           the group member id is kept in the gen_server looping state and
+%%           it is reused when re-joining the group.
+%%  - rejoin_delay_seconds (optional, default = 1)
+%%      Delay in seconds before re-joining the group.
+%%  - offset_commit_policy (optional, default = commit_to_kafka_v2)
+%%      How/where to commit offsets, possible values:
+%%        - commit_to_kafka_v2:
+%%            Group controller will commit the offsets to kafka using
+%%            version 2 OffsetCommitRequest.
+%%        - consumer_managed:
+%%            The subscirber (brodh_group_subscriber.erl) is responsible
+%%            for persisting offsets to a local or centralized storage.
+%%            And the callback get_committed_offsets should be implemented
+%%            to allow group controller to retrieve the commited offsets.
+%%  - offset_commit_interval_seconds (optional, default = 60)
```

Comment:
5 seconds to kafka --- new java client
60 seconds to zookeeper -- old java client